go/pkg/casng/batching.go

```diff
@@ -314,12 +318,113 @@ func (u *uploader) writeBytes(ctx context.Context, name string, r io.Reader, siz
 //
 // This method must not be called after cancelling the uploader's context.
 func (u *BatchingUploader) Upload(ctx context.Context, reqs ...UploadRequest) ([]digest.Digest, Stats, error) {
-	panic("not yet implemented")
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload; reqs=%d", len(reqs))
+	defer contextmd.Infof(ctx, log.Level(1), "[casng] upload.done; reqs=%d", len(reqs))
+
+	var stats Stats
+
+	if len(reqs) == 0 {
+		return nil, stats, nil
+	}
+
+	var undigested []UploadRequest
+	digested := make(map[digest.Digest]UploadRequest)
+	var digests []digest.Digest
+	for _, r := range reqs {
+		if r.Digest.Hash == "" {
+			undigested = append(undigested, r)
+			continue
+		}
+		digested[r.Digest] = r
+		digests = append(digests, r.Digest)
+	}
+	missing, err := u.MissingBlobs(ctx, digests)
+	if err != nil {
+		return nil, stats, err
+	}
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload; missing=%d, undigested=%d", len(missing), len(undigested))
+
+	reqs = undigested
+	for _, d := range missing {
+		reqs = append(reqs, digested[d])
+		delete(digested, d)
+	}
+	for d := range digested {
+		stats.BytesRequested += d.Size
+		stats.LogicalBytesCached += d.Size
+		stats.CacheHitCount += 1
+		stats.DigestCount += 1
+	}
+	if len(reqs) == 0 {
+		contextmd.Infof(ctx, log.Level(1), "[casng] upload: nothing is missing")
+		return nil, stats, nil
+	}
+
+	contextmd.Infof(ctx, log.Level(1), "[casng] upload: uploading; reqs=%d", len(reqs))
+	ch := make(chan UploadRequest)
+	resCh := u.streamPipe(ctx, ch)
+
+	u.clientSenderWg.Add(1)
+	go func() {
+		defer close(ch) // let the streamer terminate.
+		defer u.clientSenderWg.Done()
+
+		contextmd.Infof(ctx, log.Level(1), "[casng] upload.sender.start")
+		defer contextmd.Infof(ctx, log.Level(1), "[casng] upload.sender.stop")
+
+		for _, r := range reqs {
+			select {
+			case ch <- r:
+			case <-u.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	var uploaded []digest.Digest
+	for r := range resCh {
+		if r.Err != nil {
+			err = errors.Join(r.Err, err)
```

Comment:
There's no streamPipe implementation yet, so I'm not sure whether it's doable or even make sense, but wouldn't it be better to return an error as soon as we got one instead of processing the rest of the requests?

My concern is, if RBE starts to timeout (e.g. because it's not reachable or overloaded, or because of misconfigured endpoint) we might end up waiting `request_timeout * len(reqs)` to get the response instead of `request_timeout` if we returned the error just after one failed request.

Reply:
The requests go through multiple workers downstream. Such errors would be somewhere in the middle and by the time the request arrives here, potentially multiple requests would already be within the pipeline. If the server is not responding leaving the connection to timeout, that would be considered a retryable error. Right now the retry logic is per-call (may be multiple calls per request), which is not too bad.

If a broadcast cancellation is desired, that could be implemented using the streaming interface of this package which I'm going to submit later. Using that interface, the user gets responses as soon as they are received without internal batching, allowing the user to cancel the call and abort in-flight requests.